File: playwright-java/src/test/java/testtracks/configuration/EnvironmentTargets.java
package testtracks.configuration;

import java.util.Map;

public record EnvironmentTargets(
    Map<String, SiteTarget> sites,
    Map<String, ApiTarget> apis,
    Map<String, CredentialTarget> credentials,
    EnvironmentRunSettings settings) {

  public SiteTarget site(String name) {
    return target(sites, name, "site");
  }

  public ApiTarget api(String name) {
    return target(apis, name, "api");
  }

  public CredentialTarget credential(String name) {
    return target(credentials, name, "credential");
  }

  private static <T> T target(Map<String, T> values, String name, String type) {
    var value = values.get(name);
    if (value == null) {
      throw new IllegalStateException(
          "The " + type + " target '" + name + "' was not found in all-targets.json.");
    }
    return value;
  }
}

File: playwright-csharp/tests/Configuration/EnvironmentTargets.cs
namespace TestTracks.Playwright.CSharp.Specs.Configuration;

public sealed record EnvironmentTargets(
    IReadOnlyDictionary<string, SiteTarget> Sites,
    IReadOnlyDictionary<string, ApiTarget> Apis,
    IReadOnlyDictionary<string, CredentialTarget> Credentials,
    EnvironmentRunSettings Settings)
{
    public SiteTarget Site(string name) => GetTarget(Sites, name, "site");

    public ApiTarget Api(string name) => GetTarget(Apis, name, "api");

    public CredentialTarget Credential(string name) => GetTarget(Credentials, name, "credential");

    private static T GetTarget<T>(IReadOnlyDictionary<string, T> values, string name, string type)
    {
        return values.TryGetValue(name, out var value)
            ? value
            : throw new InvalidOperationException($"The {type} target '{name}' was not found in all-targets.json.");
    }
}

public sealed record SiteTarget(string BaseUrl);

public sealed record ApiTarget(string BaseUrl);

public sealed record CredentialTarget(string Username, string Password);

public sealed record EnvironmentRunSettings(
    int DefaultTimeoutMs,
    int ExpectTimeoutMs,
    bool TraceOnFailure);
